pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,212 @@
+package alertmanager
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/alertmanager/cluster"
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// state represents the Alertmanager silences and notification log internal state
+type state struct {
+	userID string
+
+	mtx    sync.Mutex
+	states map[string]cluster.State
+	logger log.Logger
+	reg    prometheus.Registerer
+
+	replicationFactor  int
+	replicateStateFunc func(context.Context, string, *clusterpb.Part) error
+
+	partialStateMergesTotal  *prometheus.CounterVec
+	partialStateMergesFailed *prometheus.CounterVec
+	stateReplicationTotal    *prometheus.CounterVec
+	stateReplicationFailed   *prometheus.CounterVec
+
+	msgc   chan *clusterpb.Part
+	stopc  chan struct{}
+	readyc chan struct{}
+}
+
+func (s *state) Name() string {
+	return ""
```

Comment:
What's the impact of returning empty string?

Reply:
I added a comment on 

```
		// Cortex should not expose cluster information back to its tenants.
		Peer:     &state{},
```

Seems like my attempt to make it more visible was futile, I'll add one to these methods.